velocity_modelling/cvm/basin_model.py

```diff
@@ -0,0 +1,696 @@
+
+    # Superseded by determine_basin_surface_above_vectorized(), but kept for reference
+
+    # def determine_basin_surface_above(self, depth: float):
+    #     """
+    #     Determine the index of the basin surface directly above the given depth.
+    #
+    #     Parameters
+    #     ----------
+    #     depth : float
+    #         The depth of the grid point to determine the properties at.
+    #
+    #     Returns
+    #     -------
+    #     int
+    #         Index of the surface directly above the grid point.
+    #     """
+    #     # self.depths is in decreasing order
+    #
+    #     valid_indices = np.where((~np.isnan(self.depths)) & (self.depths >= depth))[0]
+    #     return valid_indices[-1] if valid_indices.size > 0 else 0  # the last one
+
+    def determine_basin_surface_above_vectorized(self, depths: np.ndarray):
+        """
+        Vectorized approach for finding the indices of the basin surfaces directly above each of the given `depths`.
+
+        Parameters
+        ----------
+        depths : np.ndarray
+            Array of depth values to test.
+
+        Returns
+        -------
+        np.ndarray
+            Indices of surfaces that are directly above each depth.
+        """
+        # Ensure depths is a NumPy array
+        depths = np.asarray(depths)
+
+        # Initialize output array with default value 0
+        indices = np.zeros_like(depths, dtype=int)
+
+        # Mask for valid depths in self.depths (not NaN)
+        valid_mask = ~np.isnan(self.depths)
+        valid_depths = self.depths[valid_mask]
+        valid_indices = np.where(valid_mask)[0]
+
+        if valid_depths.size == 0:
+            return indices  # All depths are invalid, return zeros
+
+        # For each depth, find the last valid index where self.depths >= depth
+        # Since self.depths is in decreasing order, we can use searchsorted
+        # searchsorted finds the insertion point where depth would be inserted to maintain order
+        # Since depths are decreasing, we want the rightmost index where valid_depths >= depth
+        search_indices = np.searchsorted(-valid_depths, -depths, side="right")
+
+        # Adjust indices to account for valid_mask
+        valid_search_indices = np.clip(search_indices - 1, 0, valid_depths.size - 1)
+        indices = valid_indices[valid_search_indices]
+
+        # If no valid depths are >= depth, return 0 (as in original logic)
+        indices[search_indices == 0] = 0
+
+        return indices
+
+    def enforce_basin_surface_depths(
+        self,
+        in_basin: InBasin,
+        mesh_vector: MeshVector,
+    ):
+        """
+        Enforce hierarchy of surfaces, then mark which depths lie within the basin.
+
+        Parameters
+        ----------
+        in_basin : InBasin
+            Tracks if a lat-lon is inside the basin.
+        mesh_vector : MeshVector
+            Coordinates for the point of interest.
+        """
+        assert in_basin.in_basin_lat_lon
```

Comment:
Would prefer a more explicit `raise` expression here.
```suggestion
        if not in_basin.in_basin_lat_lon:
            raise ValueError('Point is not contained in basin')
```
this should also be documented in the function.

Reply:
Note that I could be wrong about the interpretation of the error your throwing because it isn't documented what that `assert` actually checks.

Reply:
I think I can just remove it. 
This is where original C code explicitly checked if in_basin_lat_lon is true, but I already pre-processed it, and basin membership is guaranteed.

![image](https://github.com/user-attachments/assets/ae7f2a04-3cbe-4df6-8fdc-15f87daca0f8)